### ISSUE TITLE
non_max_suppression hangs for some cases

### ISSUE THREAD

**Issue by favrei** [eriklindernoren/PyTorch-YOLOv3#315]
The non_max_suppression in utils.py removes boxes by detecting "large overlap" boxes in line 252.
Later, those large overlapping boxes will be removed except for the one with highest confidence.
However, at the very beginning of training, the predicted box could be very crazy like x1y1x2y2 = (nan, nan, nan, nan) or (0, 100, 0, 150) which have no area or so called "IOU".
The non_max_suppression hangs at this point because it tries to eliminate large overlap boxes but no box is going to be overlapping with the examples above.

https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/utils/utils.py#L252-L260

**Comment by apple635471** [comment]
@killer9619 

Hi, i also encountered the same problem. 
I try to write some dirty code to deal with this problem as below.
Except for these 2 rare case you said will happen at the beginning step. I also thought of another case if the predicted box beyond the border of the image. How do you think?
The following is a very dirty code to process these abnormal case, if you have another better idea, please let me know. Thanks!!

```python=
def filter_wrong_results(prediction, image_size):

    prediction[..., :4] = xywh2xyxy(prediction[..., :4])
    output = [None for _ in range(len(prediction))]
    
    for image_i, image_pred in enumerate(prediction):

        keep_boxes = []
        for box_i, box_pred in enumerate(prediction[image_i]):
       
            if box_pred[0] < 0: box_pred[0] = 0.
            if box_pred[1] < 0: box_pred[1] = 0.
            if box_pred[2] >= image_size: box_pred[2] = image_size - 1.
            if box_pred[3] >= image_size: box_pred[3] = image_size - 1.

            if math.isnan(box_pred[0].item()): continue
            if math.isnan(box_pred[1].item()): continue
            if math.isnan(box_pred[2].item()): continue
            if math.isnan(box_pred[3].item()): continue            
            if box_pred[2] - box_pred[0] == 0: continue
            if box_pred[3] - box_pred[1] == 0: continue
            keep_boxes += [box_pred]

        if keep_boxes:
            output[image_i] = torch.stack(keep_boxes)

    return output
```

**Comment by killer9619** [comment]
@apple635471 
Hi, thanks for your input.
Here are my thoughts:
1. Points outside of the boundary will not be a problem because it is going to "large overlap" with itself as long as the area of the box is not 0. Yes, of course, it is apparently a wrong detection but still acceptable at the very beginning of the training.
2. I did something more brutal by detecting if the amount of "large overlap" bboxes is 0. When this happens, just remove the candidate box.
To be specific, I added following code after the large_overlap detection.
```
if not any(large_overlap):
    detections = np.delete(detections, 0, 0)
    continue
```
